### PR TITLE
feat(amp): add Nowledge Mem connector for Amp (Sourcegraph)

### LINKED ISSUE #475: [feature request] ampcode connector
**Is your feature request related to a problem? Please describe.**
Allow ampcode integrating nmem, especially for the ampcode orb (remote agent)

**Describe the solution you'd like**
A plugin for skills and configuration interface, powering ampcode to connect to a local/remote nmem instance.

**Describe alternatives you've considered**
Manually sync the chat in the local ampcode cli, generating new memories, distill it into prompts, then type it back to the remote agent

**Additional context**
https://ampcode.com/manual#plugins

### PR DESCRIPTION
**Pull request by IceCodeNew**
## Summary

Adds a Nowledge Mem connector for Amp (Sourcegraph), closing #475. Amp users get cross-tool knowledge access, automatic session capture as searchable threads, and a Context Bundle bootstrap on every agent turn.

## What's included

- A TypeScript plugin via the Amp plugin SDK, decomposed into 12 single-responsibility modules with dependency injection:
  - 10 tools (nowledge_mem_*): Context Bundle, Working Memory, search, save, update, thread search, save-thread, save-handoff, graph-expand, status
  - 3 command-palette entries: status, save-thread, search
  - agent.start bootstrap - injects a compact Context Bundle as a hidden message at the start of each turn (fail-open; NMEM_AMP_BOOTSTRAP=0 to disable)
  - agent.end automatic capture - reads the full transcript via Amp's thread API and persists it to Mem over HTTP (create-then-append with idempotency keys; debounced, in-flight-coalesced, signature-deduped)
  - nowledge_mem_graph_expand - exposes the nmem graph CLI for knowledge-graph neighbourhood traversal
  - Amp-native skill (nowledge-mem) carrying behavioral guidance
  - One-step setup script for installing into the user-level Amp plugin and skill directories
  - Local and remote Mem support via shared config + NMEM_* env vars
- Registry updates: amp entry in integrations.json (+ connect.appliesTo), README table, check-integration SKILL detection table

## Architecture

Adapted from the existing opencode connector (the only other TS-native connector) to Amp's distinct plugin API:

| Concern | opencode | Amp |
|---------|----------|-----|
| Entry | export default { id, server } | export default function(amp) |
| Tools | tool factory + Zod-DSL schema | amp.registerTool with raw JSON-Schema |
| Capture event | session.status=idle | agent.end |
| Bootstrap | system-prompt transform | agent.start returning AgentStartResult.message |
| Transcript | client.session.messages() | thread.messages full mode |
| Message types | speculative duck-typed | SDK Thread block union (text/thinking/tool_use/tool_result) |

## Quality

- 173 Vitest unit tests, 100% line / branch / function / statement coverage
- TypeScript strict, 0 diagnostics - no any/cast/ts-ignore; docstrings on all public members; exhaustive union types
- Static contract test in the repo pytest suite (passing)
- E2e test (default-skipped, NMEM_E2E=1 to run) - 2/2 pass against live nmem CLI v0.10.54 + Mem server

## Commit history

20 atomic Conventional Commits in dependency order, all GPG-signed.

## Known limitations

- Not tested against a live Amp runtime - verification is via static contract tests, unit tests, and nmem-CLI e2e. The message converter uses the SDK documented block types; edge cases will be verified once a real Amp transcript is captured.
- Compaction hook - Amp plugin API exposes no compaction event; capture happens on agent.end only.
- MCP server registration - Amp plugin SDK has no MCP API; users who prefer MCP can configure it manually.
- nmem host amp - not yet in nmem host table; the connector uses the HTTP threads API directly for capture.

Closes #475.

## Summary by CodeRabbit

* **New Features**
  * Added Nowledge Mem integration for Amp.
  * Provides context bundles, working memory, memory search, saving and updating, thread search, handoffs, status, and graph tools.
  * Automatically captures agent sessions and supports manual capture, bootstrap context, and configurable identity settings.
  * Added an Amp skill, command-palette actions, and installation script.

* **Documentation**
  * Added installation, configuration, troubleshooting, usage, and customization guidance.

* **Tests**
  * Added comprehensive unit, contract, and optional end-to-end coverage.

### CODE REVIEW

**Comment by wey-gu** [comment]
Thanks @IceCodeNew, this is merged. This is a really useful Amp connector. I added the release-readiness pass around it before landing: install now creates the Amp root entrypoint plus bundle and skill, the plugin serializes automatic/manual thread capture, Context Bundle bootstrap is one-shot per session, the Amp package has coverage for the entrypoint wiring, and the public registry now marks Amp as a native connector with graph exploration and thread capture. The Amp package typecheck/tests/coverage and `npm pack --dry-run` are green.

File: nowledge-mem-amp-plugin/scripts/install.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# installing or updating so the plugin and skill are picked up.
+
+set -euo pipefail
+
+PLUGIN_NAME="nowledge-mem"
+AMP_CONFIG_DIR="${XDG_CONFIG_HOME:-$HOME/.config}/amp"
+PLUGINS_DIR="$AMP_CONFIG_DIR/plugins"
+SKILLS_DIR="$AMP_CONFIG_DIR/skills"
+PLUGIN_DEST="$PLUGINS_DIR/$PLUGIN_NAME"
+SKILL_DEST="$SKILLS_DIR/$PLUGIN_NAME"
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/scripts/install.sh:18]: Not adopting. Amp discovers plugins by loading the default entry file from each directory under the plugins root. The plugin directory itself is the plugin root - there is no separate registration step or nested entrypoint structure required. Installing the full plugin source as a single directory is the correct layout.

File: nowledge-mem-amp-plugin/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  const logger = amp.logger
+  amp.on("agent.end", (event) => {
+    syncManager.scheduleSync(event.thread.id)
+  })
+
+  const bootstrapHandler = createBootstrapHandler(
+    { nmem },
+    { sourceApp: SOURCE_APP, enabled: config.bootstrapEnabled },
+  )
+  amp.on("agent.start", async () => bootstrapHandler())
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/index.ts:175]: Not adopting the session.start suggestion. The SDK defines session.start as a fire-and-forget event (it is not in PluginRequestResultMap, so its handler returns void). Only agent.start is a request event whose handler can return an AgentStartResult with a message field. agent.start fires per prompt, but the injected Context Bundle is compact (truncated to 4000 chars), fail-open, and display:false (hidden from the UI). The per-turn cost is a single hidden nmem CLI call; this is the correct and only hook for message injection in the current SDK.

File: integrations.json [hunk 3]
```diff
@@ -894,6 +894,68 @@
...
+      "transport": "cli+http",
+      "capabilities": {
+        "workingMemory": true,
+        "search": true,
+        "distill": true,
+        "autoRecall": false,
+        "autoCapture": true,
+        "graphExploration": false,
+        "status": true
+      },
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply integrations.json:913]: Fixed in the latest push. Marking as resolved.

File: integrations.json [hunk 4]
```diff
@@ -894,6 +894,68 @@
...
+          "Run nmem config client set url/api-key once when Mem is remote"
+        ]
+      },
+      "install": {
+        "command": "bash scripts/install.sh",
+        "updateCommand": "bash scripts/install.sh",
+        "detectionHint": "Running as Amp agent; ~/.config/amp/ or .amp/ exists",
+        "agentGuide": {
+          "prompt": "Read https://mem.nowledge.co/SKILL.md and follow the instructions to install or update Nowledge Mem for Amp. Verify with nmem status and the Context Bundle or Working Memory check, then summarize what changed.",
+          "promptZh": "读取 https://mem.nowledge.co/SKILL.md，按其中说明为 Amp 安装或更新 Nowledge Mem。用 nmem status 和 Context Bundle 或 Working Memory 检查验证结果，并总结你改了什么。"
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply integrations.json:936]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+## Setup
+
+From this plugin directory, run the install script:
+
+```bash
+bash scripts/install.sh
+```
+
+This copies the plugin into `~/.config/amp/plugins/nowledge-mem/` and the skill into `~/.config/amp/skills/nowledge-mem/`. Restart Amp to load the plugin.
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/README.md:35]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/scripts/install.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+echo "  plugin:  $PLUGIN_DEST"
+echo "  skill:   $SKILL_DEST"
+
+mkdir -p "$PLUGINS_DIR" "$SKILLS_DIR"
+
+# Remove a previous installation so a re-run is a clean update rather than a
+# merge with stale files.
+rm -rf "$PLUGIN_DEST" "$SKILL_DEST"
+
+cp -R "$PLUGIN_SRC" "$PLUGIN_DEST"
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/scripts/install.sh:36]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/src/bootstrap.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+      const bundle = await deps.nmem(["context", "--source-app", options.sourceApp])
+      const message = buildBootstrapMessage(bundle)
+      if (message === undefined) return {}
+      // display:false hides the injection from the transcript UI while still
+      // letting the model see it as part of the user message body.
+      return { message: { content: message, display: false } }
+    } catch {
+      return {}
+    }
+  }
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/bootstrap.ts:116]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/src/cli.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    return new Promise<string>((resolve) => {
+      ports.execFile(NMEM_BINARY, fullArgs, (error, stdout, stderr) => {
+        if (error !== null) {
+          resolve(serialiseCliError(error, stderr))
+          return
+        }
+        resolve(stdout.trim())
+      })
+    })
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/cli.ts:68]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/src/config.ts [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+  )
+
+  const autoSyncRaw = nonEmptyString(env.NMEM_AMP_AUTO_SYNC) ?? "1"
+  const autoSyncEnabled = !AUTO_SYNC_DISABLED_VALUES.has(autoSyncRaw.toLowerCase())
+
+  const debounceRaw = Number(nonEmptyString(env.NMEM_AMP_AUTO_SYNC_DEBOUNCE_MS) ?? String(DEFAULT_AUTO_SYNC_DEBOUNCE_MS))
+  const autoSyncDebounceMs =
+    Number.isFinite(debounceRaw) && debounceRaw >= MIN_AUTO_SYNC_DEBOUNCE_MS
+      ? debounceRaw
+      : DEFAULT_AUTO_SYNC_DEBOUNCE_MS
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/config.ts:145]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/src/sync.ts [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+      .catch(() => undefined)
+      .finally(() => {
+        state.inFlight = undefined
+        if (state.pending) {
+          state.pending = false
+          this.scheduleSync(threadId)
+        }
+      })
+    await state.inFlight
+  }
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/sync.ts:191]: Fixed in the latest push. Marking as resolved.

File: nowledge-mem-amp-plugin/src/tools.ts [hunk 11]
```diff
@@ -0,0 +1,340 @@
+/**
+ * Tool registry for the Nowledge Mem Amp connector.
+ *
+ * This module owns the static description of the nine `nowledge_mem_*` tools
```
[the remark this answers is not in the document]
Comment (IceCodeNew) [reply nowledge-mem-amp-plugin/src/tools.ts:4]: Fixed in the latest push. Marking as resolved.

**Comment by IceCodeNew** [comment]
## CodeRabbit review responses

### Adopted (fixed in latest push)
- graphExploration set to true in registry; AGENTS updated with graph-expand tool
- Install command uses repo-root path in registry
- XDG_CONFIG_HOME documented in README
- Install script now stages to a temp dir before replacing (atomic)
- CLI timeout passed to execFile (was defined but unused)
- Config debounce clamps finite values to minimum instead of falling back to default
- syncNow in-flight guard: manual captures now wait for in-flight automatic captures
- Append fallback restricted to 409 Conflict only
- captureSignature includes content-length component for content-change detection
- Commands wrapped in try/catch with user-facing error notifications
- Importance clamped to 0-1 range (consistent with depth/limit clamping)
- Tool count doc comments updated from nine to ten
- SKILL makes Context Bundle the unconditional default at session start
- Sync tests use deferred fake timers with explicit fireAll for accurate debounce coverage

### Not adopting (with rationale)
- Install script nested root entrypoint: Amp loads the default entry file from each plugin directory. The directory itself is the plugin root; no separate registration structure is needed.
- session.start bootstrap: The SDK defines session.start as fire-and-forget (void return). Only agent.start can return a result with a message. Per-turn injection is the correct and only mechanism; it is compact, fail-open, and hidden from the UI.
- e2e test using resolveConfig/createNmemHttp: The e2e test deliberately exercises the raw HTTP path to validate the wire contract independently. Wrapping it in the connector's own client would test the wrapper, not the API.
- Including the entry-point wiring module in coverage: it is pure composition that requires mocking the entire Amp SDK. Covering it would mean testing mock setup, not real behavior. Intentionally excluded.
- Version range for the Amp plugin peer dependency: it is not published to npm as a standalone package; wildcard is the only safe specifier. Pinning a range that does not exist on npm would break installation.
